### PR TITLE
Listen on mouseup for when a user click the up qty field buttons

### LINKED ISSUE #985: Apple Pay not initially showing correct total (quantity) if used on Product Page
#### Affected ticket(s)

2332304-zen

#### What I expected

When the Apple Pay button is used from an individual product page, the initial total should take into acccount the product's price X the quantity selected (ex: `$15.00 x 2`).

#### What happened instead

The checkout totals if Apple Pay is used from the product page are incorrect, as they don't take quantity into account for the total initially, they only show the product price 1X: 

https://cld.wthms.co/j9jlrW
Image link: https://cld.wthms.co/j9jlrW

Then click button: 

https://cld.wthms.co/B0wXnr
Image link: https://cld.wthms.co/B0wXnr

As we can see, it should show $30 (`$15 x quantity of 2`), but only shows $15.00 - it also doesn't display the total correctly if there is a `Default` variation option set.

The end total that gets calculated is correct though.

#### Steps to reproduce the issue

1. Setup Apple Pay.
2. Setup a variable product.
3. Go to the individual product page on the front end and select a variation option with a quantity of 2 or more.
4. When selecting `Apple Pay`, the initial total does not reflect the quantity selected, although the end total is correct.

Not sure if this is a Stripe/Apple Pay limitation, but either way it's `Low` priority as the final total is correct.

-------------------

- [ ] Issue assigned to next milestone.
- [ ] Issue assigned a priority (will be assessed by maintainers).

### PR DESCRIPTION
**Pull request by dwainm**
Fixes #985

#### Changes proposed in this Pull Request:
- In addition to keyup, also listen on mouseup.

### Testing
Not easy to test but if you have apple pay/or other payment request methods working:

On the single product page click the up qty buttons and see if the value reflects in Apple Pay.
-------------------
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

**Comment by vbelolapotkov** [comment]
I've experimented with this noticed the following:

1. Once the payment button clicked: displayed items shows product qty and cost without shipping. Once shipping info loaded, product qty is not shown anymore, only shipping cost and total. I.e. product qty and cost are shown in transition only. 
2. Debouncing is missing for the qty event handler. Clicking it multiple times (long enough) leads to qty mismatch in the UI and payment request.

Those two seem to be different issues, not sure if we need to address them here and now.
@dwainm what do you think? If we're not going to handle those issues now, it's safe to ship the PR.

**Comment by dwainm** [comment]
@vbelolapotkov I'm not sure about 1. That feels like a separate issue and probably expected if no shipping info is loaded. 

For the second issue, you raised. I agree. I'll look into adding a fix for that.

**Comment by vbelolapotkov** [comment]
> @vbelolapotkov I'm not sure about 1. That feels like a separate issue and probably expected if no shipping info is loaded.

I agree, it's a separate issue. IMO it's rather about items cost are not shown anymore after shipping info loaded. I'd expect items, shipping, tax and total are specified once ready to process payment.

> For the second issue, you raised. I agree. I'll look into adding a fix for that.

Thanks!

File: assets/js/stripe-payment-request.js [hunk 1]
```diff
@@ -520,7 +535,7 @@ jQuery( function( $ ) {
 				});
 			} );
 
-			$( '.quantity' ).on( 'keyup', '.qty', function () {
+			$( '.quantity' ).on( 'keyup mouseup', '.qty', wc_stripe_payment_request.debounce( 250, function() {
```
Comment (RadoslavGeorgiev) [inline assets/js/stripe-payment-request.js:538]: In the context of `mouseup` and `keyup`, I think you can simply use the `change` event. It works immediately both in Chrome and Safari.

File: assets/js/stripe-payment-request.js [hunk 2]
```diff
@@ -424,6 +424,21 @@ jQuery( function( $ ) {
 			} );
 		},
 
+		debounce: function ( wait, func, immediate ) {
```
Comment (RadoslavGeorgiev) [inline assets/js/stripe-payment-request.js:427]: This is probably not needed, but could you please lint/format this snippet?

Comment (dwainm) [reply assets/js/stripe-payment-request.js:427]: @RadoslavGeorgiev I tried linting and got no issues.Also, this feels like it should be part of the build script?

File: assets/js/stripe-payment-request.js (hunk 1 above)
Comment (dwainm) [reply assets/js/stripe-payment-request.js:538]: I'm not going be able to test these again before the next sprint so leaving it as is.

File: assets/js/stripe-payment-request.js (hunk 2 above)
Comment (jrodger) [inline assets/js/stripe-payment-request.js:427]: Would it help to comment this function? The meaning of "debounce" may not be immediately obvious to someone.

Comment (dwainm) [reply assets/js/stripe-payment-request.js:427]: Sorry, I had an issue with linter tried again and it's giving me lots of error across the whole file. Most of them really meaningless and not really sure this plugin is set up correctly for formatting and listing... I don't mind doing it but it's going to take a while to fix this all up. Going to defer it to another issue:
https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1209

**Comment by jrodger** [comment]
Catching the mouseup is definitely an improvement, it shuts down one avenue where the quantity in the field and the payment button can get out of sync (i.e. clicking the number up/down buttons). The only downside I can see is the payment button refreshing if you click (or tap on mobile) the quantity field, without changing anything.

Using the `change` event instead, as @RadoslavGeorgiev suggested might resolve that, although I haven't investigated whether there is a reason that wasn't used in the first place.

There are still ways to get the 2 values out of sync however, perhaps we should open new issue for that?

One way to de-sync things:

* Highlight the quantity
* Hover your mouse over the "Buy now" button
* Press a number on your keyboard, and quickly click the button.

This feels like a bit of a losing battle, it's tricky because Stripe injects the button for us, but as a future fix I wonder if we could add some sort of click handler wrapper on the button that grabs the quantity rather than trying to trap every possible way that the quantity can be updated.

File: assets/js/stripe-payment-request.js (hunk 2 above)
Comment (dwainm) [reply assets/js/stripe-payment-request.js]: Added in 0a9b2f6

File: assets/js/stripe-payment-request.js (hunk 1 above)
Comment (dwainm) [reply assets/js/stripe-payment-request.js]: Sorry again, had a moment to do a quick test and this works nicely. Fixed up in: 
e41344d

**Comment by luizreis** [comment]
I tested the `change` event locally, and it looks like it's not fired right after a keyboard input but rather when the element loses focus. That could lead to customers editing the value manually and clicking the button right away. The arrow controls work, but we should probably use the `input` event to fit both of those cases.

Also, if customers click the button before the quantity change is propagated, Stripe throws the following error in the console, and checkout is blocked until the page is refreshed:

`Uncaught IntegrationError: You cannot update Payment Request options while the payment sheet is showing.`. 

This can be reproduced by typing in a quantity and clicking the button without taking focus out of the quantity input. 

IMO, if using debouncing, we should probably figure out a way to block the payment button before the debounce event runs. It could be as simple as adding another listener (without debouncing) and blocking the button on that.

```
$( '.quantity' ).on( 'input', '.qty', function() {
	wc_stripe_payment_request.blockPaymentRequestButton( prButton );
} );
```

cc @dwainm

**Comment by dwainm** [comment]
New PR here: 
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1211

**Pull request merged**